Sources/StreamChat/Repositories/AuthenticationRepository.swift

```diff
@@ -140,6 +153,98 @@ class AuthenticationRepository {
         scheduleTokenFetch(userInfo: nil, tokenProvider: tokenProviderCheckingSuccess, completion: completion)
     }
 
+    func prepareEnvironment(
+        userInfo: UserInfo?,
+        newToken: Token,
+        completion: @escaping (Error?) -> Void
+    ) {
+        let state: EnvironmentState
+        if currentUserId == nil {
+            state = .firstConnection
+        } else if newToken.userId == currentUserId {
+            state = .newToken
+        } else {
+            state = .newUser
+        }
+
+        log.assert(delegate != nil, "Delegate should not be nil at this point")
+
+        switch state {
+        case .firstConnection:
+            connectionRepository.updateWebSocketEndpoint(with: newToken, userInfo: userInfo)
+            setToken(token: newToken, completeTokenWaiters: true)
+            delegate?.didFinishSettingUpAuthenticationEnvironment(for: state)
+            completion(nil)
+
+        case .newToken:
+            connectionRepository.updateWebSocketEndpoint(with: newToken, userInfo: userInfo)
+            setToken(token: newToken, completeTokenWaiters: true)
+            delegate?.didFinishSettingUpAuthenticationEnvironment(for: state)
+            completion(nil)
+
+        case .newUser:
+            completeTokenWaiters(token: nil)
+            setToken(token: newToken, completeTokenWaiters: false)
+            delegate?.didFinishSettingUpAuthenticationEnvironment(for: state)
+
+            // Setting a new connection is not possible in connectionless mode.
+            guard connectionRepository.isClientInActiveMode else {
+                completion(ClientError.ClientIsNotInActiveMode())
+                return
+            }
+
+            connectionRepository.disconnect(source: .userInitiated) { [weak delegate, weak connectionRepository] in
+                connectionRepository?.updateWebSocketEndpoint(with: newToken, userInfo: userInfo)
```

Comment:
I think I've commented somewhere about the order of these two - it's clearer if we first clear current user data and then update the connection repository with a new token.

Reply:
It looks semantically incorrect. The second method has a completion block, and we should not make assumptions on whether the completion block will be run sync or async. That's why I kept that order.

Another option would be to add this line to the completion of `clearCurrentUserData`, but seems too much for this.

Reply:
ok, let's keep it in this order then.